shotover-proxy/src/codec/cassandra.rs

```diff
@@ -52,60 +55,65 @@ impl CassandraCodec {
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>> {
         // if we need to close the connection return an error.
         if let Some(result) = self.force_close.take() {
             debug!("Closing errored connection: {:?}", &result);
             return Err(anyhow!(result));
         }
 
-        // TODO: We could implement our own version and length check here directly on the bytes to avoid the duplicate frame parse
-        match RawCassandraFrame::from_buffer(src, self.compressor) {
-            Ok(parsed_frame) => {
-                // Clear the read bytes from the FramedReader
-                let bytes = src.split_to(parsed_frame.frame_len);
-                tracing::debug!(
-                    "incoming cassandra message:\n{}",
-                    pretty_hex::pretty_hex(&bytes)
-                );
+        loop {
+            // TODO: We could implement our own version and length check here directly on the bytes to avoid the duplicate frame parse
+            match RawCassandraFrame::from_buffer(src, self.compressor) {
+                Ok(parsed_frame) => {
+                    // Clear the read bytes from the FramedReader
+                    let bytes = src.split_to(parsed_frame.frame_len);
+                    tracing::debug!(
+                        "incoming cassandra message:\n{}",
+                        pretty_hex::pretty_hex(&bytes)
+                    );
 
-                Ok(Some(vec![Message::from_bytes(
-                    bytes.freeze(),
-                    MessageType::Cassandra,
-                )]))
-            }
-            Err(ParseFrameError::NotEnoughBytes) => Ok(None),
-            Err(ParseFrameError::UnsupportedVersion(version)) => {
-                // if we got an error force the close on the next read.
-                // We can not immediately close as we are gong to queue a message
-                // back to the client and we have to allow time for the message
-                // to be sent.  We can not reuse the connection as it may/does contain excess
-                // data from the failed parse.
-                self.force_close = Some(format!(
-                    "Received frame with unknown protocol version: {}",
-                    version
-                ));
+                    self.messages
+                        .push(Message::from_bytes(bytes.freeze(), MessageType::Cassandra));
+                }
+                Err(ParseFrameError::NotEnoughBytes) => {
+                    if self.messages.is_empty() || src.remaining() != 0 {
+                        return Ok(None);
+                    } else {
+                        return Ok(Some(std::mem::take(&mut self.messages)));
+                    }
+                }
+                Err(ParseFrameError::UnsupportedVersion(version)) => {
+                    let _ = src.split_to(9); // we will have read the entire header so clear those bytes
```

Comment:
I dont see the point in partially handling this here?
We have to force close the connection anyway as happens on the following line.

Reply:
We still have to clear the buffer so the loop can exit.

Reply:
Clearing just the header isnt enough.
I guess you could clear the entire buffer but I think it would be more clear if you just returned from the function.